File: backend/app/schemas/drug_events.py
from pydantic import BaseModel

from app.schemas.audit import AuditSummary


class DrugEventReaction(BaseModel):
    reaction: str
    count: int


class DrugSignalIntelligenceScore(BaseModel):
    score: int
    label: str
    data_confidence: str
    top_reaction_concentration: float
    review_priority: str
    score_version: str
    limitations: list[str]


class DrugReactionCategory(BaseModel):
    category: str
    count: int
    reactions: list[str]


class DrugSignalTrendSnapshot(BaseModel):
    label: str
    current_record_count: int
    previous_record_count: int | None = None
    previous_audit_id: str | None = None
    previous_created_at: str | None = None
    explanation: str
    limitation: str
    trend_version: str

class DrugSignalSemanticMatch(BaseModel):
    record_id: str
    text: str
    similarity_score: float
    explanation: str
    source_name: str | None = None


class DrugSignalSemanticPreview(BaseModel):
    query_text: str
    matches: list[DrugSignalSemanticMatch]
    limitations: list[str]
    preview_version: str
    is_production_ml: bool


class DrugEventSearchResponse(BaseModel):
    query: str
    count: int
    limit: int
    source_name: str
    endpoint: str
    retrieval_timestamp: str
    medical_disclaimer: str
    faers_disclaimer: str
    audit: AuditSummary
    intelligence_score: DrugSignalIntelligenceScore
    reaction_categories: list[DrugReactionCategory]
    reaction_classifier_version: str
    trend_snapshot: DrugSignalTrendSnapshot
    top_reactions: list[DrugEventReaction]

File: frontend/src/api/drugEvents.ts
import { apiClient } from './client'

export type AuditSummary = {
  audit_id: string
  source_id: string
  module: string
  upstream_status: string
  record_count: number
  transform_version: string
}

export type DrugEventSort = 'reports' | 'alpha'

export type DrugEventReaction = {
  reaction: string
  count: number
}

export type DrugSignalIntelligenceScore = {
  score: number
  label: string
  data_confidence: string
  top_reaction_concentration: number
  review_priority: string
  score_version: string
  limitations: string[]
}

export type DrugReactionCategory = {
  category: string
  count: number
  reactions: string[]
}

export type DrugSignalTrendSnapshot = {
  label: string
  current_record_count: number
  previous_record_count: number | null
  previous_audit_id: string | null
  previous_created_at: string | null
  explanation: string
  limitation: string
  trend_version: string
}

export type DrugEventSearchResponse = {
  query: string
  raw_query?: string
  normalized_query?: string
  correction_applied?: boolean
  suggestion_message?: string | null
  count: number
  limit: number
  source_name: string
  endpoint: string
  retrieval_timestamp: string
  medical_disclaimer: string
  faers_disclaimer: string
  audit: AuditSummary
  intelligence_score: DrugSignalIntelligenceScore
  reaction_categories: DrugReactionCategory[]
  reaction_classifier_version: string
  trend_snapshot: DrugSignalTrendSnapshot
  top_reactions: DrugEventReaction[]
}

export async function searchDrugEvents(query: string, limit = 10, sort: DrugEventSort = 'reports') {
  const response = await apiClient.get<DrugEventSearchResponse>(
    `/api/v1/drug-events/search`,
    {
      params: {
        q: query,
        limit,
        sort,
      },
    }
  )

  return response.data
}
